File: app/src/main/java/com/ladsoft/bakingapp/adapter/StepSlideshowAdapter.java
package com.ladsoft.bakingapp.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;

import com.ladsoft.bakingapp.entity.Step;
import com.ladsoft.bakingapp.fragment.RecipeStepFragment;

import java.util.ArrayList;
import java.util.List;

public class StepSlideshowAdapter extends FragmentStatePagerAdapter {
    private final RecipeStepFragment.Callback listener;
    private List<Step> steps = new ArrayList<>();

    public StepSlideshowAdapter(FragmentManager fragmentManager, RecipeStepFragment.Callback listener) {
        super(fragmentManager);
        this.listener = listener;
    }

    @Override
    public Fragment getItem(int position)  {
        RecipeStepFragment fragment = RecipeStepFragment.newInstance();
        fragment.setListener(listener);

        return fragment;
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }

    @Override
    public int getCount() {
        return steps.size();
    }

    public void setDataSource(List<Step> steps) {
        this.steps = steps == null
                ? new ArrayList<Step>()
                : steps;

        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/adapter/StepSlideshowAdapter.kt
package com.ladsoft.bakingapp.adapter

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import com.ladsoft.bakingapp.entity.Step
import com.ladsoft.bakingapp.fragment.RecipeStepFragment


class StepSlideshowAdapter(val fragmentManager: FragmentManager, val listener: RecipeStepFragment.Callback): FragmentStatePagerAdapter(fragmentManager) {
    private var steps : List<Step> = mutableListOf()

    override fun getItem(position: Int): Fragment {
        return RecipeStepFragment.newInstance().apply{
            setListener(listener)
        }
    }

    override fun getItemPosition(`object`: Any): Int {
        return POSITION_NONE
    }

    override fun getCount(): Int = steps.size

    fun setDataSource(steps: List<Step>?) {
        when {
            steps == null -> this.steps = mutableListOf()
            else -> this.steps = steps
        }

        notifyDataSetChanged()
    }
}